hivemind/server/expert_backend.py

```diff
@@ -121,18 +135,53 @@ def backward(self, *inputs: torch.Tensor) -> Tuple[torch.Tensor, ...]:
                 nested_flatten(grad_outputs), outputs_flat))
             torch.autograd.backward(outputs_flat, grad_tensors=grad_outputs_flat,
                                     create_graph=False, retain_graph=False)
-            self.apply_gradients()
+            self.apply_gradients(batch_size)
 
         return tuple(x.grad if isinstance(x.grad, torch.Tensor) else torch.zeros_like(x)
                      for x in nested_flatten((args, kwargs)))
 
-    def apply_gradients(self) -> None:
+    def apply_gradients(self, batch_size) -> None:
         """
         Train the expert for one step. This method is called by ``ExpertBackend.backward`` after computing gradients.
         """
-        self.opt.step()
-        self.opt.zero_grad()
+        self.optimizer.step()
+        self.optimizer.zero_grad()
+
+        if self.scheduler is not None:
+            self.scheduler.step()
+
         self.update_count += 1
+        self.examples_processed += batch_size
+
+    def get_stats(self) -> Dict:
+        """
+        Return current expert training statistics (number of updates, number of processed examples after last optimizer step)
+        """
+        return {
+            'updates': self.update_count,
+            'examples_processed': self.examples_processed
+        }
+
+    def get_full_state(self) -> Dict:
+        """
+        Return the current state of the expert (including batch processing statistics)
+        """
+        full_state = {
+            'stats': self.get_stats(),
+            'model': self.expert.state_dict(),
+            'optimizer': self.optimizer.state_dict(),
+            'scheduler': {} if self.scheduler is None else self.scheduler.state_dict()
+        }
+        return full_state
+
+    def load_full_state(self, state_dict: Dict):
+        self.update_count = state_dict['stats']['updates']
+        self.examples_processed = state_dict['stats']['examples_processed']
+
+        self.expert.load_state_dict(state_dict['model'])
+        self.optimizer.load_state_dict(state_dict['optimizer'])
+        if self.scheduler is not None:
+            self.scheduler.load_state_dict(state_dict['scheduler'])
```

Comment:
I think that existence of keys "optimizer", "stats" and "scheduler" in the state_dict should be checked before loading_state. This might be helpful if the user has already trained his own model and wants to load it as an expert.

Reply:
Done